js/flexbuffers.js

```diff
@@ -0,0 +1,1015 @@
+  }
+
+  function intStackValue(value) {
+    return StackValue(flexbuffers.ValueType.INT, flexbuffers.BitWidth.width(value), value);
+  }
+
+  function uintStackValue(value) {
+    return StackValue(flexbuffers.ValueType.UINT, flexbuffers.BitWidth.uwidth(value), value);
+  }
+
+  function floatStackValue(value) {
+    return StackValue(flexbuffers.ValueType.FLOAT, flexbuffers.BitWidth.fwidth(value), value);
+  }
+
+  function offsetStackValue(offset, valueType, bitWidth) {
+    return StackValue(valueType, bitWidth, null, offset);
+  }
+
+  function finishBuffer() {
+    if (stack.length !== 1) {
+      throw `Stack has to be exactly 1, but it is ${stack.length}. You have to end all started vectors and maps before calling [finish]`;
+    }
+    const value = stack[0];
+    const byteWidth = align(value.elementWidth(offset, 0));
+    writeStackValue(value, byteWidth);
+    writeInt(value.storedPackedType(), 1);
+    writeInt(byteWidth, 1);
+    finished = true;
+  }
+
+  return  {
+    add: function (value) {
+      integrityCheckOnValueAddition();
+      if (typeof value === 'undefined') {
+        throw "You need to provide a value";
+      }
+      if (value === null) {
+        stack.push(nullStackValue());
+      } else if (typeof value === "boolean") {
+        stack.push(boolStackValue(value));
+      } else if (typeof value == 'number') {
+        if (Number.isInteger(value)) {
+          stack.push(intStackValue(value));
+        } else {
+          stack.push(floatStackValue(value));
+        }
+      } else if (ArrayBuffer.isView(value)){
+        writeBlob(value.buffer);
+      } else if (typeof value === 'string' || value instanceof String) {
+        writeString(value);
+      } else if (Array.isArray(value)) {
+        startVector();
+        for (let i = 0; i < value.length; i++) {
+          this.add(value[i]);
+        }
+        end();
+      } else if (typeof value === 'object'){
+        const properties = Object.getOwnPropertyNames(value).sort();
+        startMap(true);
+        for (let i = 0; i < properties.length; i++) {
+          const key = properties[i];
+          this.addKey(key);
+          this.add(value[key]);
+        }
+        end();
+      } else {
+        throw `Unexpected value input ${value}`;
+      }
+    },
+    finish: function() {
+      if (!finished) {
+        finishBuffer();
+      }
+      const result = buffer.slice(0, offset);
+      return new Uint8Array(result);
+    },
+    isFinished: function() {
+      return finished;
+    },
+    addKey: function(key) {
+      integrityCheckOnKeyAddition();
+      writeKey(key);
+    },
+    addInt: function(value, indirect = false, cache = false) {
+      integrityCheckOnValueAddition();
+      if (!indirect) {
+        stack.push(intStackValue(value));
+        return;
+      }
+      if (cache && indirectIntCache.hasOwnProperty(value)) {
+        stack.push(indirectIntCache[value]);
```

Comment:
there's a cache for indirect int values? why?

Reply:
It's based on following API:

`addInt: function(value, indirect = false, cache = false)`

If I want to add an int indirectly, I can also cache it so if I add another int with the same value it will be reused. I think there is similar strategy in C++ but there user need to store the offset and provide it. It is faster but not as safe as users might provide a wrong offset. As JS is in general a language where it is easier to make mistakes ;) I chose to provide an API which provides smaller surface for errors.

Reply:
Yes, but these lookups are expensive and potentially use a lot of memory. I'd say they should be optional, but in the case of indirect ints I think not having them at all is better.

The FlatBuffers JS API also doesn't have any check for wrong offsets, it is a very high cost to pay to try and make this foolproof.

Reply:
Yeah agree, this is also way the deduplication strategy is set do false per default. To be honest I don't thin that many users will use a builder in JS, I assume >90% users will just call `encode` with a dict or array.